libpromises/mod_custom.c

```diff
@@ -0,0 +1,741 @@
+
+    JsonDestroy(response);
+
+    if (!valid)
+    {
+        // Detailed error messages from module should already have been printed
+        const char *const promise_type = pp->parent_section->promise_type;
+        const char *const promiser = pp->promiser;
+        const char *const filename =
+            pp->parent_section->parent_bundle->source_path;
+        const size_t line = pp->offset.line;
+        Log(LOG_LEVEL_ERR,
+            "%s:%zu: %s promise with promiser '%s' failed validation",
+            filename,
+            line,
+            promise_type,
+            promiser);
+    }
+
+    return valid;
+}
+
+static inline const char *LogLevelToRequestFromModule()
+{
+    // We will never request LOG_LEVEL_NOTHING or LOG_LEVEL_CRIT from the
+    // module:
+    const LogLevel global = LogGetGlobalLevel();
+    if (global < LOG_LEVEL_ERR)
+    {
+        assert((global == LOG_LEVEL_NOTHING) || (global == LOG_LEVEL_CRIT));
+        return LogLevelToString(LOG_LEVEL_ERR);
+    }
+    return LogLevelToString(global);
+}
+
+static PromiseResult PromiseModule_Evaluate(
+    PromiseModule *module, EvalContext *ctx, const Promise *pp)
+{
+    assert(module != NULL);
+    assert(pp != NULL);
+
+    PromiseModule_AppendString(module, "operation", "evaluate_promise");
+    PromiseModule_AppendString(
+        module, "log_level", LogLevelToRequestFromModule());
+    PromiseModule_AppendString(module, "promiser", pp->promiser);
+
+    PromiseModule_AppendAllAttributes(module, pp);
+    PromiseModule_Send(module);
+
+    JsonElement *response = PromiseModule_Receive(module);
+    PromiseResult result;
+    const char *const result_str = JsonObjectGetAsString(response, "result");
+
+    Attributes a = GetClassContextAttributes(ctx, pp); // TODO: WTF
+
+    if (result_str == NULL)
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_ERR,
+            result,
+            pp,
+            &a,
+            "Promise module did not return a result for promise evaluation (bug in promise module: %s)",
+            module->path);
+    }
+    else if (StringEqual(result_str, "kept"))
+    {
+        result = PROMISE_RESULT_NOOP;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was kept by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "not_kept"))
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was not kept by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "repaired"))
+    {
+        result = PROMISE_RESULT_CHANGE;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
```

Comment:
This should be `LOG_LEVEL_INFO`, AFAICT.

Reply:
The promise module provides the INFO messages, they are already printed at this point.